src/rpc/common/Checkers.hpp

```diff
@@ -0,0 +1,134 @@
+    ANY  SPECIAL,  DIRECT,  INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#pragma once
+
+#include "rpc/Errors.hpp"
+#include "rpc/common/ValidationHelpers.hpp"
+
+#include <boost/json/value.hpp>
+#include <boost/json/value_to.hpp>
+#include <fmt/core.h>
+
+#include <optional>
+#include <string>
+#include <string_view>
+#include <utility>
+#include <vector>
+
+namespace rpc::check {
+/**
+ * @brief Warning that checks can return
+ */
+struct Warning {
+    /**
+     * @brief Construct a new Warning object
+     *
+     * @param code The warning code
+     * @param message The warning message
+     */
+    Warning(WarningCode code, std::string message) : warningCode(code), extraMessage(std::move(message))
+    {
+    }
+
+    bool
+    operator==(Warning const& other) const = default;
+
+    WarningCode warningCode;
+    std::string extraMessage;
+};
+using Warnings = std::vector<Warning>;
+
+/**
+ * @brief Check for a deprecated fields
+ */
+template <typename... T>
+class Deprecated;
+
+/**
+ * @brief Check if a field is deprecated
+ */
+template <>
+class Deprecated<> final {
+public:
+    /**
+     * @brief Check if a field is deprecated
+     *
+     * @param value The json value to check
+     * @param key The key to check
+     * @return A warning if the field is deprecated or std::nullopt otherwise
+     */
+    [[nodiscard]] static std::optional<Warning>
+    check(boost::json::value const& value, std::string_view key)
+    {
+        if (value.is_object() and value.as_object().contains(key))
+            return Warning{WarningCode::warnRPC_DEPRECATED, fmt::format("Field '{}' is deprecated.", key)};
+        return std::nullopt;
+    }
+};
+
+/**
+ * @brief Check if a value of a field is deprecated
+ * @tparam T The type of the field
+ */
+template <typename T>
+class Deprecated<T> final {
+    T value_;
+
+public:
+    /**
+     * @brief Construct a new Deprecated object
+     *
+     * @param val The value that is deprecated
+     */
+    Deprecated(T val) : value_(val)
+    {
+    }
+
+    /**
+     * @brief Check if a value of a field is deprecated
+     *
+     * @param value The json value to check
+     * @param key The key to check
+     * @return A warning if the field is deprecated or std::nullopt otherwise
+     */
+    [[nodiscard]] std::optional<Warning>
+    check(boost::json::value const& value, std::string_view key) const
+    {
+        if (value.is_object() and value.as_object().contains(key) and
+            validation::checkType<T>(value.as_object().at(key))) {
+            using boost::json::value_to;
+            auto const res = value_to<T>(value.as_object().at(key));
```

Comment:
Does it throw error? If I have a check Deprecated(2), but field value is "ok"?

Reply:
There is a type check in the if above. If the field has a different type it will be ignored.

Reply:
Ha I see. For example, if there is a type restrict Type<std::string> ahead, but the field value is int. Does Clio send back the "deprecated" warning? or no warning?

Reply:
If we specify `{"some_field", Deprecated{"some string"}}` but in the request it is `"some_field": 1` there will be no warning because deprecated value only string "some string".